File: java/soup.java
import soup.*;
import vendor1.Vendor1;


interface Controller {
  Device getDevice();

  default void foo(int param) {
    getDevice().op1(param * param);
  }
}


class ControllerA implements Controller {
  DeviceA dev;

  public ControllerA(DeviceA dev) {
    this.dev = dev;
  }

  public DeviceA getDevice() { return dev; }

  int run(int param) {
    return dev.op2(param);
  }
}


class ControllerB implements Controller {
  DeviceB dev;

  public ControllerB(DeviceB dev) {
    this.dev = dev;
  }

  public DeviceB getDevice() { return dev; }

  int run(int param) {
    return dev.op2(param, param);
  }
}


class Runner {
  public int runDevices(int init1, int init2, int param) {
    Vendor vendor1 = new Vendor1();

    DeviceA devA = vendor1.createDeviceA(1);
    DeviceB devB = vendor1.createDeviceB(2);

    ControllerA ctlA = new ControllerA(devA);
    ControllerB ctlB = new ControllerB(devB);

    ctlA.foo(init1);
    ctlB.foo(init2);
    return ctlB.run( ctlA.run(param) );
  }
}


class soup {

  public static void main(String[] args) {

    int res = new Runner().runDevices(13, 15, 42);
    System.out.println(res);
  }
}

File: scala/soup.scala


object soup {

    trait Device {
        def op1(param: Int)
    }

    trait DeviceA extends Device {
        def op2(param: Int): Int
    }

    trait DeviceB extends Device {
        def op2(x: Int, y: Int): Int
    }

    trait Vendor {
        def createDeviceA(index: Int): DeviceA
        def createDeviceB(index: Int): DeviceB
    }
}


object control {
    import soup._

    trait Controller[T <: Device] {
        val dev: T
        def foo(param: Int) {
            dev.op1(param * param)
        }
    }

    class ControllerA(val dev: DeviceA) extends Controller[DeviceA] {
        def run(param: Int): Int = {
            dev.op2(param)
        }
    }

    class ControllerB(val dev: DeviceB) extends Controller[DeviceB] {
        def run(param: Int): Int = {
            dev.op2(param, param)
        }
    }
}


object vendor1 {
    import soup._

    class Vendor1DeviceA(var x: Int) extends DeviceA {
        def op1(param: Int): Unit = { x = param }
        def op2(param: Int): Int = { x += 1; x }
    }

    class Vendor1DeviceB(var x: Int) extends DeviceB {
        def op1(param: Int): Unit = { x *= param }
        def op2(x: Int, y: Int): Int = { this.x * x + y }
    }

    class Vendor1 extends Vendor {
        def createDeviceA(index: Int): DeviceA = new Vendor1DeviceA(index)
        def createDeviceB(index: Int): DeviceB = new Vendor1DeviceB(index)
    }
}



object runner {
    import control._


    def run_devices(init1: Int, init2: Int, param: Int): Int = {
        val v1 = new vendor1.Vendor1()
        val devA = v1.createDeviceA(1)
        val devB = v1.createDeviceB(2)

        val ctlA = new ControllerA(devA)
        val ctlB = new ControllerB(devB)

        ctlA.foo(init1)
        ctlB.foo(init2)
        ctlB.run( ctlA.run(param) )
    }

    def main(args: Array[String]) = {
        val x = runner.run_devices(13, 15, 42)
        println(x)
    }
}
